### ISSUE TITLE
IterableDataset with batch samples of different size

### ISSUE THREAD

**Issue by Slyne** [huggingface/accelerate#348]
Hi,
I want to train a large dataset with accelerate.
However, my dataset has different shape (same batch, different sequence length) of samples.
This [line](https://github.com/huggingface/accelerate/blob/b2956acbe925845bf202224797334cfe4da18396/src/accelerate/data_loader.py#L356) requires the big batch for different processes to maintain a same shape so that it can concat.

I was wondering if we can just broadcast the tensors to other processors without forcing the data to be of same shape?

**Comment by sgugger** [comment]
Yes, that's a current limitation of the `dispatch_batches=True` option. I will try to dive into this and find a better solution for batches of different sizes, but that's going to take a bit of time.

**Comment by npuichigo** [comment]
Any update on this? @sgugger by the way, what's the advantage of using dispatch_batches?

**Comment by JohnHerry** [comment]
the same problem, Why should IterableDataset keep all  MAX_LEN  sample to make batches? it is a waste of memory.
